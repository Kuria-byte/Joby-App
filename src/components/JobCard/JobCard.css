.job-card {
  width: 90%; /* Occupies 90% of the screen width */
  max-width: 800px; /* Optional: Set a max width */
  margin: 20px auto; /* Center the card */
  padding: 20px; /* Add some padding */
  border-radius: 16px; /* Rounded corners */
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); /* Subtle box-shadow */
  background-color: white; /* Ensure background is white */
  transition: box-shadow 0.3s ease; /* Smooth transition for hover effects */
}

.job-card:hover {
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2); /* Slightly increase shadow on hover */
}

.job-card-header {
  display: flex;
  align-items: center;
  margin-bottom: 20px;
}

.company-image {
  width: 64px;
  height: 64px;
  margin-right: 16px;
  border-radius: 12px;
  overflow: hidden;
  background: #f5f5f5;
}

.company-image img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.company-info {
  flex: 1;
}

.company-name {
  font-size: 1.1rem;
  color: #2c3e50;
  margin: 0 0 8px 0;
}

.job-meta {
  display: flex;
  gap: 16px;
  font-size: 0.9rem;
  color: #64748b;
}

.location, .job-type {
  display: flex;
  align-items: center;
  gap: 6px;
}

.job-card-body {
  margin-bottom: 20px;
}

.job-title {
  font-size: 1.5rem;
  color: #1a202c;
  margin: 0 0 12px 0;
  line-height: 1.3;
}

.salary-badge {
  display: inline-flex;
  align-items: center;
  gap: 6px;
  background: #4CAF50;
  color: white;
  padding: 6px 12px;
  border-radius: 20px;
  font-size: 0.95rem;
  margin-bottom: 16px;
}

.job-description {
  color: #4a5568;
  line-height: 1.6;
  margin: 0;
  font-size: 0.95rem;
}

.job-card-footer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 20px;
  padding-top: 20px;
  border-top: 1px solid #e2e8f0;
}

.posted-date {
  display: flex;
  align-items: center;
  gap: 6px;
  color: #64748b;
  font-size: 0.9rem;
}

.action-buttons {
  display: flex;
  gap: 12px;
}

.action-buttons button {
  display: flex;
  align-items: center;
  gap: 8px;
  padding: 8px 16px;
  border-radius: 8px;
  font-size: 0.95rem;
  font-weight: 500;
  cursor: pointer;
  transition: all 0.2s ease;
}

.btn-more-info {
  background: #f1f5f9;
  color: #475569;
  border: none;
}

.btn-more-info:hover {
  background: #e2e8f0;
}

.btn-apply {
  background: #4CAF50;
  color: white;
  border: none;
}

.btn-apply:hover {
  background: #45a049;
}

/* Responsive Design */
@media (max-width: 480px) {
  .job-card {
    padding: 16px;
    margin: 12px;
  }

  .company-image {
    width: 48px;
    height: 48px;
  }

  .job-title {
    font-size: 1.3rem;
  }

  .action-buttons {
    flex-direction: column;
    width: 100%;
  }

  .action-buttons button {
    width: 100%;
    justify-content: center;
  }

  .job-meta {
    flex-direction: column;
    gap: 8px;
  }
}
