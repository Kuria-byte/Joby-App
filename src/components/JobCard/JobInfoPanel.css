.job-info-panel {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: white;
  padding: 20px;
  border-top-left-radius: 16px;
  border-top-right-radius: 16px;
  box-shadow: 0 -2px 10px rgba(0, 0, 0, 0.2);
  transition: transform 0.3s ease;
}

.close-button {
  background: none;
  border: none;
  font-size: 24px;
  cursor: pointer;
}

.apply-button {
  background-color: green;
  color: white;
  padding: 10px 20px;
  border: none;
  border-radius: 8px;
  cursor: pointer;
}
