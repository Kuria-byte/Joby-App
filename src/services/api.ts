const API_BASE_URL = process.env.REACT_APP_API_BASE_URL || 'http://localhost:3001';

export interface Job {
  id: string;
  title: string;
  company: string;
  location: string;
  description: string;
}

export const api = {
  async getJobs(): Promise<Job[]> {
    const response = await fetch(`${API_BASE_URL}/jobs`);
    if (!response.ok) {
      throw new Error('Failed to fetch jobs');
    }
    return response.json();
  },

  async likeJob(jobId: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/jobs/${jobId}/like`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    if (!response.ok) {
      throw new Error('Failed to like job');
    }
  },

  async passJob(jobId: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/jobs/${jobId}/pass`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    if (!response.ok) {
      throw new Error('Failed to pass job');
    }
  },
};
