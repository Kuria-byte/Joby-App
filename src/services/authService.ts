import axios from 'axios';

export interface LoginCredentials {
  email: string;
  password: string;
}

export interface SignupCredentials extends LoginCredentials {
  name: string;
  confirmPassword: string;
}

export interface AuthResponse {
  token: string;
  user: {
    id: string;
    name: string;
    email: string;
  };
}

class AuthService {
  private baseURL = '/api/auth';
  private tokenKey = 'joby_auth_token';

  async login(credentials: LoginCredentials): Promise<AuthResponse> {
    try {
      const response = await axios.post(`${this.baseURL}/login`, credentials);
      this.setToken(response.data.token);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 401) {
          throw new Error('Invalid email or password');
        }
        if (error.response?.status === 429) {
          throw new Error('Too many login attempts. Please try again later');
        }
      }
      throw new Error('Login failed. Please try again');
    }
  }

  async signup(credentials: SignupCredentials): Promise<AuthResponse> {
    if (credentials.password !== credentials.confirmPassword) {
      throw new Error('Passwords do not match');
    }

    try {
      const response = await axios.post(`${this.baseURL}/signup`, credentials);
      this.setToken(response.data.token);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 409) {
          throw new Error('Email already exists');
        }
        if (error.response?.status === 400) {
          throw new Error('Invalid signup information');
        }
      }
      throw new Error('Signup failed. Please try again');
    }
  }

  async logout(): Promise<void> {
    try {
      await axios.post(`${this.baseURL}/logout`);
    } finally {
      this.removeToken();
    }
  }

  async validateToken(): Promise<boolean> {
    const token = this.getToken();
    if (!token) return false;

    try {
      await axios.get(`${this.baseURL}/validate`, {
        headers: { Authorization: `Bearer ${token}` }
      });
      return true;
    } catch {
      this.removeToken();
      return false;
    }
  }

  getToken(): string | null {
    return localStorage.getItem(this.tokenKey);
  }

  private setToken(token: string): void {
    localStorage.setItem(this.tokenKey, token);
  }

  private removeToken(): void {
    localStorage.removeItem(this.tokenKey);
  }
}

export const authService = new AuthService();
