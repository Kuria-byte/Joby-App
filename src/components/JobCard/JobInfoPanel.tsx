import React from 'react';
import './JobInfoPanel.css'; // Import CSS for styling

interface JobInfoPanelProps {
  job: {
    title: string;
    company: string;
    description: string;
    requirements: string;
    location: string;
    salary: string;
  };
  onClose: () => void;
}

const JobInfoPanel: React.FC<JobInfoPanelProps> = ({ job, onClose }) => {
  return (
    <div className="job-info-panel">
      <button className="close-button" onClick={onClose}>X</button>
      <h2>{job.title}</h2>
      <h3>{job.company}</h3>
      <p><strong>Location:</strong> {job.location}</p>
      <p><strong>Salary:</strong> {job.salary}</p>
      <p><strong>Description:</strong> {job.description}</p>
      <p><strong>Requirements:</strong> {job.requirements}</p>
      <button className="apply-button">Apply Now</button>
    </div>
  );
};

export default JobInfoPanel;
