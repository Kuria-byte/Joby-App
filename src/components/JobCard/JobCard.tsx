import React from 'react';
import './JobCard.css';
import { Job } from '../../services/api';

interface JobCardProps {
  job: Job;
}

const JobCard: React.FC<JobCardProps> = ({ job }) => {
  const {
    title,
    company,
    location,
    description,
    salary = 'Salary not specified',
    imageUrl,
    postedDate,
    jobType
  } = job;

  const defaultImage = '/icons/company-default.svg';

  return (
    <div className="job-card">
      <div className="job-card-header">
        <div className="company-image">
          <img 
            src={imageUrl || defaultImage} 
            alt={`${company} logo`}
            onError={(e) => {
              const target = e.target as HTMLImageElement;
              target.src = defaultImage;
            }}
          />
        </div>
        <div className="company-info">
          <h3 className="company-name">{company}</h3>
          <div className="job-meta">
            <span className="location">
              <i className="fas fa-map-marker-alt"></i> {location}
            </span>
            {jobType && (
              <span className="job-type">
                <i className="fas fa-briefcase"></i> {jobType}
              </span>
            )}
          </div>
        </div>
      </div>

      <div className="job-card-body">
        <h2 className="job-title">{title}</h2>
        <div className="salary-badge">
          <i className="fas fa-money-bill-wave"></i> {salary}
        </div>
        <p className="job-description">{description}</p>
      </div>

      <div className="job-card-footer">
        {postedDate && (
          <span className="posted-date">
            <i className="far fa-clock"></i> Posted {postedDate}
          </span>
        )}
        <div className="action-buttons">
          <button className="btn-more-info">
            <i className="fas fa-info-circle"></i>
            More Info
          </button>
          <button className="btn-apply">
            <i className="fas fa-paper-plane"></i>
            Apply Now
          </button>
        </div>
      </div>
    </div>
  );
};

export default JobCard;
