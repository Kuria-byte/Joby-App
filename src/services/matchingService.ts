import { Job } from './api';

export interface Match {
  id: string;
  jobId: string;
  userId: string;
  employerId: string;
  jobTitle: string;
  companyName: string;
  matchedAt: string;
  status: 'pending' | 'active' | 'closed';
}

export interface MatchResponse {
  isMatch: boolean;
  match?: Match;
}

class MatchingService {
  private socket: WebSocket;

  constructor() {
    this.socket = new WebSocket('ws://localhost:3000/matches');
    this.setupSocketListeners();
  }

  private setupSocketListeners() {
    this.socket.onmessage = (event) => {
      const data = JSON.parse(event.data);
      if (data.type === 'match') {
        this.handleNewMatch(data.match);
      }
    };
  }

  private handleNewMatch(match: Match) {
    // Dispatch a custom event that components can listen to
    const matchEvent = new CustomEvent('newMatch', { detail: match });
    window.dispatchEvent(matchEvent);
  }

  async checkForMatch(userId: string, jobId: string): Promise<MatchResponse> {
    try {
      const response = await fetch('http://localhost:3000/api/matches/check', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ userId, jobId }),
      });
      
      if (!response.ok) {
        throw new Error('Failed to check for match');
      }

      return await response.json();
    } catch (error) {
      console.error('Error checking for match:', error);
      throw error;
    }
  }

  async getMatches(userId: string): Promise<Match[]> {
    try {
      const response = await fetch(`http://localhost:3000/api/matches/user/${userId}`);
      if (!response.ok) {
        throw new Error('Failed to fetch matches');
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching matches:', error);
      throw error;
    }
  }
}

export const matchingService = new MatchingService();
