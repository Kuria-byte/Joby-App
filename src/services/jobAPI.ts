import axios from 'axios';
import { Job } from './api';
import { matchingService, MatchResponse } from './matchingService';

const API_BASE_URL = 'http://localhost:3000/api'; // Change this to your actual API base URL

export interface JobInterestData {
  jobId: string;
  userId: string;
  action: 'interested' | 'dismissed';
  timestamp: string;
}

export const jobAPI = {
  // Record user's interest in a job
  recordJobInterest: async (jobId: string, userId: string): Promise<MatchResponse> => {
    try {
      const data: JobInterestData = {
        jobId,
        userId,
        action: 'interested',
        timestamp: new Date().toISOString(),
      };
      
      await axios.post(`${API_BASE_URL}/jobs/interest`, data);
      
      // Check for a match after recording interest
      return await matchingService.checkForMatch(userId, jobId);
    } catch (error) {
      console.error('Error recording job interest:', error);
      throw error;
    }
  },

  // Record job dismissal
  dismissJob: async (jobId: string, userId: string): Promise<void> => {
    try {
      const data: JobInterestData = {
        jobId,
        userId,
        action: 'dismissed',
        timestamp: new Date().toISOString(),
      };
      
      await axios.post(`${API_BASE_URL}/jobs/dismiss`, data);
    } catch (error) {
      console.error('Error dismissing job:', error);
      throw error;
    }
  },

  // Get job recommendations
  getJobRecommendations: async (userId: string, limit: number = 10): Promise<Job[]> => {
    try {
      const response = await axios.get<Job[]>(`${API_BASE_URL}/jobs/recommendations`, {
        params: { userId, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching job recommendations:', error);
      throw error;
    }
  }
};
