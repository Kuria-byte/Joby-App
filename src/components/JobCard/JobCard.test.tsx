import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import JobCard from './JobCard';
import { Job } from '../../services/api';


const mockOnInfo = jest.fn();

const mockJob: Job = {
  id: '1',
  title: 'Software Engineer',
  company: 'Tech Corp',
  location: 'San Francisco, CA',
  description: 'Exciting role for a software engineer',
  salary: '$120,000 - $150,000',
  imageUrl: 'https://example.com/logo.png',
  postedDate: '2025-02-11',
  jobType: 'Full-time',
  requirements: 'Experience with JavaScript and React.'
};

describe('JobCard', () => {
  it('renders job information correctly', () => {
    const mockOnLogout = jest.fn();
    render(<JobCard job={mockJob} onInfo={mockOnInfo} />);
    
    expect(screen.getByText('Tech Corp')).toBeInTheDocument();
    expect(screen.getByText('San Francisco, CA')).toBeInTheDocument();
    expect(screen.getByText('Full-time')).toBeInTheDocument();
    expect(screen.getByText('$120,000 - $150,000')).toBeInTheDocument();
    expect(screen.getByAltText('Tech Corp logo')).toBeInTheDocument();
  });

  it('uses default image when company logo fails to load', () => {
    const mockOnLogout = jest.fn();
    render(<JobCard job={mockJob}  onInfo={mockOnInfo} />);
    
    const img = screen.getByAltText('Tech Corp logo') as HTMLImageElement;
    fireEvent.error(img);
    
    expect(img.src).toContain('/icons/company-default.svg');
  });

  it('renders without optional fields', () => {
    const jobWithoutOptionals: Job = {
      id: '2',
      title: 'Developer',
      company: 'Startup Inc',
      location: 'Remote',
      description: 'Great opportunity',
      salary: 'Not specified', 
      imageUrl: '',
      postedDate: '2025-02-11',
      jobType: 'Full-time' ,
      requirements: 'Experience with JavaScript and React.'
    };

    const mockOnLogout = jest.fn();
    render(<JobCard job={jobWithoutOptionals}  onInfo={mockOnInfo} />);
    
    expect(screen.getByText('Startup Inc')).toBeInTheDocument();
    expect(screen.getByText('Remote')).toBeInTheDocument();
    expect(screen.getByText('Salary not specified')).toBeInTheDocument();
    expect(screen.queryByText(/Full-time/)).not.toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const mockOnLogout = jest.fn();
    const { container } = render(<JobCard job={mockJob} onInfo={mockOnInfo} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
