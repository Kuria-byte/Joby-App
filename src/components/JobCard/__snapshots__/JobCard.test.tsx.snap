// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCard matches snapshot 1`] = `
<div
  class="job-card"
>
  <div
    class="job-card-header"
  >
    <div
      class="company-image"
    >
      <img
        alt="Tech Corp logo"
        src="https://example.com/logo.png"
      />
    </div>
    <div
      class="company-info"
    >
      <h3
        class="company-name"
      >
        Tech Corp
      </h3>
      <div
        class="job-meta"
      >
        <span
          class="location"
        >
          <i
            class="fas fa-map-marker-alt"
          />
           
          San Francisco, CA
        </span>
        <span
          class="job-type"
        >
          <i
            class="fas fa-briefcase"
          />
           
          Full-time
        </span>
      </div>
    </div>
  </div>
  <div
    class="job-card-body"
  >
    <h2
      class="job-title"
    >
      Software Engineer
    </h2>
    <div
      class="salary-badge"
    >
      <i
        class="fas fa-money-bill-wave"
      />
       
      $120,000 - $150,000
    </div>
    <p
      class="job-description"
    >
      Exciting role for a software engineer
    </p>
  </div>
  <div
    class="job-card-footer"
  >
    <span
      class="posted-date"
    >
      <i
        class="far fa-clock"
      />
       Posted 
      2025-02-11
    </span>
    <div
      class="action-buttons"
    >
      <button
        class="btn-more-info"
      >
        <i
          class="fas fa-info-circle"
        />
        More Info
      </button>
      <button
        class="btn-apply"
      >
        <i
          class="fas fa-paper-plane"
        />
        Apply Now
      </button>
    </div>
  </div>
</div>
`;
